### ISSUE TITLE
Cloud Controller Manager Beta Requirements

### ISSUE THREAD

**Issue by luxas** [kubernetes/kubernetes#48690]
Forked from https://github.com/kubernetes/features/issues/88 in order to allow more discussion.

Things needed for beta
-----------------------------------

1. Lots of docs (tutorials, blogs, walkthroughs etc)
2. Testing with various clouds (dogfooding)
3. Automatically build and create a image for cloud-controller-manager(CCM) on gcr.io
4. Automatically run e2e tests for CCM 
5. bugfix for kubelet certificate ip addresses (discussed here - https://github.com/kubernetes/kubernetes/pull/47152#issuecomment-307836402)
6. Make *ByID methods for all cloud integrations: https://github.com/kubernetes/kubernetes/issues/47257
7. Make possible to use an Initializer instead of the PersistentVolumeLabel admission plugin for out-of-tree clouds?

Nice to haves:
----------------------------
1. Examples of out-of-tree cloud providers (DigitalOcean, Oracle, Rancher, etc)
2. ...

Very much help wanted.
cc @wlan0 @thockin @roberthbailey @rrati @dims @andrewsykim @munnerz

**Comment by joshwget** [comment]
We can certainly help out with (1) and (2). I just updated the Rancher cloud-controller-manager to support Kubernetes 1.7 and we'll be running completely upstream with our next release as a result.

Once we test out this version and make something available I'd be happy to write up a blog post on it. The Kubernetes docs already reference https://github.com/rancher/rancher-cloud-controller-manager as a reference so we can keep this available as a public reference point.

**Comment by joshwget** [comment]
On another note, there was some talk in [a PR](https://github.com/kubernetes/kubernetes/pull/40160) that supporting out-of-tree credential providers might be in scope for cloud-controller-manager. I'm not sure if this would necessarily be a beta feature, but it would be great to determine whether this should be considered in scope or not. If not I'll file a separate issue on this since we're pretty interested in seeing this feature!

**Comment by luxas** [comment]
>  that supporting out-of-tree credential providers might be in scope for cloud-controller-manager

I definitely think that we should look into that. Seems like credential providers are only called from the node though, so it might be harder to implement... @wlan0 thoughts on that?

**Comment by wlan0** [comment]
@luxas I reached out to @erictune about this a few months back. We haven't really progressed on this much however.

It is possible to do a external credential provider implementation just like the external cloud provider implementation. 

> I definitely think that we should look into that. Seems like credential providers are only called from the node though, so it might be harder to implement... @wlan0 thoughts on that?

It depends on how we decide to implement the external credential provider. If we use a asynchronous plugin model (like for cloud-providers), it'll be tricky. 

We could resort to using a pluggable binary and a synchronous model, like with CNI plugins, then the state machine co-ordination will be simpler.

**Comment by luxas** [comment]
@wlan0 How important is it? Can it be done via a DaemonSet?

**Comment by wlan0** [comment]
@luxas, It's possible via a daemon set. But we should talk about this in more detail. 

Its not important for the cloud provider changes. It's a separate feature.

**Comment by luxas** [comment]
> Its not important for the cloud provider changes. It's a separate feature.

Agreed.

**Comment by joshwget** [comment]
Should we move this discussion to a separate Github issue then?

**Comment by luxas** [comment]
@joshwget Feel free to spawn a new issue for that

**Comment by nicksardo** [comment]
+ @cheftako

**Comment by jdumars** [comment]
/sig azure
Adding for visibility only

**Comment by luxas** [comment]
Probably makes sense to label this with all cloud related SIGs...

**Comment by andrewsykim** [comment]
> 7. Make possible to use an Initializer instead of the PersistentVolumeLabel admission plugin for out-of-tree clouds?

I know we're waiting on CSI to become a thing, but how feasible is it to support volume controller in CCM only to create the volumes and attach them via the cloud provider? If you use flex volumes, it's possible to do the agent level logic via extensions as long as the volumes are created/attached already. 

cc @luxas @wlan0

**Comment by andrewsykim** [comment]
@luxas just open sourced https://github.com/digitalocean/digitalocean-cloud-controller-manager, hoping it'll serve as a good example for cloud controller managers. It's currently in alpha, would love feedback :).

**Comment by jhorwit2** [comment]
Should kubernetes/kubernetes#49308 be considered a blocker for beta? It's pretty important to be able to reference what zone a node is in.

**Comment by luxas** [comment]
@jhorwit2 Yes, I think so. We need to sync up as a group on this though, let's see if we can find a Zoom time for interested parties...

**Comment by prydie** [comment]
We should probably also resolve https://github.com/kubernetes/kubernetes/issues/49836 pre-beta.

**Comment by jhorwit2** [comment]
Agreed. We wouldn't be able to use the provider id zone solution without first being able to set the provider id. Right now those provider id methods are basically a noop (rancher's example doesn't use them).

**Comment by luxas** [comment]
@jhorwit2 Can you contribute that feature? We need help here...

**Comment by andrewsykim** [comment]
I noticed we don't handle the case of existing nodes setting `--cloud-provider=external`. What's the plan going forward with that? Will it be the responsibility of the cluster admin to delete/create nodes or do we plan to add some logic in the kubelet to handle this?

**Comment by FengyunPan** [comment]
/sig openstack
Adding for visibility only

**Comment by tamalsaha** [comment]
#49836 is a blocker for us. I am trying to use my custom out-of-tree cloud provider with Kubernetes. But I noticed that kubelet calls apis like `NodeAddressByProviderID()` with empty `providerID` for `external` cloud provider. Digging in the code for 1.7 branch, I see that `Kubelet.Cloud` is null for `external` cloud-providers. As a result `GetInstanceProviderID(..)` returns empty string.

**Comment by jhorwit2** [comment]
@tamalsaha does the `--provider-id` flag on the kubelet help you?

**Comment by tamalsaha** [comment]
I have not tried with `--provider-id` flag. I am going to check and report back. One issue with using that flag is this is another thing that needs to happen outside kubeadm .

**Comment by tamalsaha** [comment]
We have noticed another issue related to using `external` cloud providers used with kubeadm provisioned clusters. External cloud controllers usually need working DNS to call cloud provider apis in `ByProviderID()` apis. Now, If I set `--cloud-provider=external` to Kubelet on master, I see that kube-dns pods don't run.

**Comment by jhorwit2** [comment]
What's the plan with [`alpha.kubernetes.io/provided-node-ip`](https://github.com/kubernetes/kubernetes/blob/822e29dd3ce90421c099f6bec7f476c419c3211e/pkg/kubelet/apis/well_known_annotations.go#L24) annotation for beta? Is the plan to just keep it an annotation and promote it? If so, I can create a ticket to track that as well.

**Comment by luxas** [comment]
@wlan0 @thockin ^

@jhorwit2 I don't remember the history there exactly, only vaguely

**Comment by jhorwit2** [comment]
@luxas could you please update 5 in your original post to reference https://github.com/kubernetes/kubernetes/issues/48690

**Comment by luxas** [comment]
/remove-lifecycle stale
cc @andrewsykim for an update on this effort?

**Comment by andrewsykim** [comment]
Yup, we'll be beta in 1.11 as per https://github.com/kubernetes/features/issues/88. 

The only thing we decided to leave out from the original beta requirements was E2E testing. What we would like is for all providers to run conformance tests using their CCM and report those results to testgrid but we have decided that that is not a requirement to push the CCM feature to beta since we don't see the actual APIs/interfaces changing because of it. Also trying to get all providers to publish E2E tests would take too much time and we don't want to block this feature forever :)
